sirepo/package_data/static/js/sirepo-plotting-vtk.js

```diff
@@ -2767,30 +2767,60 @@ SIREPO.app.service('vtkToPNG', function(panelState, plotToPNG, utilities) {
-                    canvas.height = parseInt(canvas3d.getAttribute('height'));
-                    if (doTraverse) {
-                        vtkRenderer.getApiSpecificRenderWindow().traverseAllPasses();
-                    }
-                    else {
-                        vtkRenderer.getRenderWindow().render();
-                    }
-                    canvas.getContext('2d').drawImage(canvas3d, 0, 0, canvas.width, canvas.height);
-                    if (axesCanvas) {
-                        canvas.getContext('2d').drawSvg(axesCanvas.outerHTML, 0, 0, canvas.width, canvas.height);
-                    }
-                });
+            copyCanvas: (event, doTraverse) => {
+                if(canvasWorking) {
+                    canvasWorking = new Object();
+                    doNextCanvasProcess = () => res.copyCanvas(event, doTraverse);
+                    return;
+                }
+                const processCanvas = () => {
+                    panelState.waitForUI(() => {
+                        const canvas3d = $(panel).find('canvas')[0];
+                        const axesCanvas = $(panel).find('svg.sr-vtk-axes')[0];
+                        canvas.width = parseInt(canvas3d.getAttribute('width'));
+                        canvas.height = parseInt(canvas3d.getAttribute('height'));
+                        if (doTraverse) {
+                            vtkRenderer.getApiSpecificRenderWindow().traverseAllPasses();
+                        }
+                        else {
+                            vtkRenderer.getRenderWindow().render();
+                        }
+                        context.drawImage(canvas3d, 0, 0, canvas.width, canvas.height);
+                        if (axesCanvas) {
+                            html2canvas(axesCanvas.parentElement,
+                            {
+                                backgroundColor: null
+                            }).then(c => {
+                                context.drawImage(c, 0, 0);
+                                canvasWorking = null;
+                                if(doNextCanvasProcess) {
+                                    const proc = doNextCanvasProcess;
+                                    doNextCanvasProcess = null;
+                                    proc();
+                                }
+                            })
+                        }
+                    });
+                }
+                const waitForUserInteractionPause = () => {
+                    canvasWorking = new Object();
+                    const tempCanvasWorking = canvasWorking;
+                    setTimeout(() => {
+                        if(tempCanvasWorking === canvasWorking) {
+                            processCanvas();
+                        } else {
+                            waitForUserInteractionPause();
+                        }
+                    }, 250);
+                }
```

Comment:
We have a [debounce](https://github.com/radiasoft/sirepo/blob/fbc1e6473dde18ffbfb2c884ea8b6154c75ca794/sirepo/package_data/static/js/sirepo-components.js#L4773) function that may serve here. Worth a look if it simplifies this.

Reply:
[c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)